### PR TITLE
core[patch]: Adds support for custom events

### PR DESCRIPTION
**Pull request by jacoblee93**

### CODE REVIEW

File: langchain-core/src/callbacks/base.ts [hunk 1]
```diff
@@ -259,6 +260,19 @@ abstract class BaseCallbackHandlerMethodsClass {
     tags?: string[]
   ): // eslint-disable-next-line @typescript-eslint/no-explicit-any
   Promise<any> | any;
+
+  handleCustomEvent?(
+    eventName: string,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    payload: any,
```
Comment (eyurtsev) [inline langchain-core/src/callbacks/base.ts:267]: `payload` vs. `data`?

File: langchain-core/src/callbacks/base.ts [hunk 2]
```diff
@@ -259,6 +260,19 @@ abstract class BaseCallbackHandlerMethodsClass {
...
+    runId: string,
+    parentRunId?: string,
+    tags?: string[],
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    metadata?: Record<string, any>,
+    name?: string
```
Comment (eyurtsev) [inline langchain-core/src/callbacks/base.ts:273]: What is name?

File: langchain-core/src/callbacks/base.ts [hunk 3]
```diff
@@ -259,6 +260,19 @@ abstract class BaseCallbackHandlerMethodsClass {
     tags?: string[]
   ): // eslint-disable-next-line @typescript-eslint/no-explicit-any
   Promise<any> | any;
+
+  handleCustomEvent?(
+    eventName: string,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    payload: any,
+    runId: string,
+    parentRunId?: string,
```
Comment (eyurtsev) [inline langchain-core/src/callbacks/base.ts:269]: Drop parentRunId b/c user can already define it, but we actually don't provide the correct one

File: langchain-core/src/callbacks/dispatch/web.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        "inside a tool or a RunnableLambda).",
+        `\n\nIf you continue to see this error, please import from "@langchain/core/callbacks/dispatch/web"`,
+        "and explicitly pass in a config parameter.",
+      ].join(" ")
+    );
+  }
+  await callbackManager.handleCustomEvent?.(
+    name,
+    payload,
+    parentRunId,
```
Comment (eyurtsev) [inline langchain-core/src/callbacks/dispatch/web.ts:62]:
```suggestion
    parentRunId, // Add a doc-string here to say it's being passed as the run id
```

File: langchain-core/src/callbacks/manager.ts [hunk 5]
```diff
@@ -860,6 +860,46 @@ export class CallbackManager
     );
   }
 
+  async handleCustomEvent?(
+    eventName: string,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    payload: any,
+    runId: string,
+    _parentRunId?: string,
```
Comment (eyurtsev) [inline langchain-core/src/callbacks/manager.ts:868]: Remove the `parentRunId` on the dispatch?

File: langchain-core/src/callbacks/base.ts (hunk 2 above)
Comment (jacoblee93) [reply langchain-core/src/callbacks/base.ts:273]: `runName`

File: langchain-core/src/callbacks/base.ts (hunk 1 above)
Comment (jacoblee93) [reply langchain-core/src/callbacks/base.ts:267]: These are just positional args, will show in tooltips but not affect anything else.

Comment (jacoblee93) [reply langchain-core/src/callbacks/base.ts:267]: Will update anyway

**Pull request merged**